File: BACKEND/admin/services/userService.js
const adminRepository = require("../models/adminRepository");
const { revokeAllSessionsForUser, isClerkConfigured } = require("../../shared/clerk/clerkClient");
const { createServiceError } = require("./errors");

const allowedRoles = new Set(["admin", "staff", "customer"]);

function normalizeText(value) {
  return String(value || "").trim().toLowerCase();
}

async function listUsers(filters = {}) {
  return adminRepository.listUsers(filters);
}

async function getUserById(id) {
  const user = await adminRepository.getUserById(id);
  if (!user) {
    throw createServiceError(404, "User account not found.");
  }
  return user;
}

function validateRoleUpdate(payload) {
  if (payload.role !== undefined && !allowedRoles.has(normalizeText(payload.role))) {
    throw createServiceError(422, "Role must be admin, staff, or customer.");
  }
}

async function updateRole(id, role) {
  validateRoleUpdate({ role });
  const user = await getUserById(id);

  const updated = await adminRepository.updateUser(id, { role });
  if (!updated) {
    throw createServiceError(404, "User account not found.");
  }
  return updated;
}

async function setUserActive(id, isActive) {
  const user = await getUserById(id);
  const targetStatus = isActive ? "active" : "locked";

  const updated = await adminRepository.updateUser(id, { status: targetStatus });
  if (!updated) {
    throw createServiceError(404, "User account not found.");
  }

  // Khi khoa: revoke session tren Clerk de dang xuat ngay lap tuc.
  let clerkResult = { revoked: 0, skipped: !isClerkConfigured() };
  if (!isActive) {
    try {
      clerkResult = await revokeAllSessionsForUser(id);
    } catch (error) {
      console.warn(`Khong the revoke session Clerk cho ${id}:`, error.message);
      clerkResult = { revoked: 0, skipped: true, error: error.message };
    }
  }

  return { ...updated, clerk: clerkResult };
}

module.exports = {
  listUsers,
  getUserById,
  updateRole,
  setUserActive,
};

File: BACKEND/admin/Services/UserService.cs
using MyPuppy.Admin.Models;
using MyPuppy.Admin.Repositories;

namespace MyPuppy.Admin.Services;

public class UserService
{
    private static readonly HashSet<string> AllowedRoles = new(StringComparer.OrdinalIgnoreCase)
    {
        "admin", "staff", "customer"
    };

    private readonly AdminRepository _repository;
    private readonly ClerkClient _clerkClient;

    public UserService(AdminRepository repository, ClerkClient clerkClient)
    {
        _repository = repository;
        _clerkClient = clerkClient;
    }

    public Task<IEnumerable<UserDto>> ListUsersAsync(string? role, string? status, string? search)
        => _repository.ListUsersAsync(role, status, search);

    public async Task<UserDto> GetUserByIdAsync(string id)
    {
        var user = await _repository.GetUserByIdAsync(id);
        if (user == null) throw new ServiceException(404, "User account not found.");
        return user;
    }

    public async Task<UserDto> UpdateRoleAsync(string id, string role)
    {
        if (string.IsNullOrWhiteSpace(role) || !AllowedRoles.Contains(role))
            throw new ServiceException(422, "Role must be admin, staff, or customer.");

        await GetUserByIdAsync(id);
        var updated = await _repository.UpdateUserAsync(id, role: role);
        return updated ?? throw new ServiceException(404, "User account not found.");
    }

    public async Task<(UserDto User, int Revoked, bool Skipped, string? Error)> SetUserActiveAsync(string id, bool isActive)
    {
        await GetUserByIdAsync(id);
        var status = isActive ? "active" : "locked";
        var updated = await _repository.UpdateUserAsync(id, status: status);
        if (updated == null) throw new ServiceException(404, "User account not found.");

        if (isActive)
            return (updated, 0, !_clerkClient.IsConfigured, null);

        var (revoked, skipped, error) = await _clerkClient.RevokeAllSessionsForUserAsync(id);
        return (updated, revoked, skipped, error);
    }
}
